### PR TITLE
Feature: Refine filters

### PR DESCRIPTION
**Pull request by KHwong12**
# Summary
This branch refines the filter panel by using alternative buttons and reorder the filters.

# Changes
The changes made in this PR are:
1. Change UI of severity type filter (`severity_filter`) to multiple checkboxes

    The old filter is a drop-down list. Now replaced with a filter of multiple checkboxes.

    

1. Add a filter of casualty type

    This filter filters the dataset by the types of casualties involved in each accident, including **Pedestrian**, **Driver** and **Passenger**. Note that selecting multiple inputs means filtering accidents including ANY of the selected classes, instead of includes ALL of the selected classes. For example, when the users choose both **Pedestrian** and **Driver** in this filter, collisions which 1. casualties include only pedestrians; 2. casualties include only drivers and 3. casualties include both pedestrians and drivers will be displayed on the map. 

       

1. Remove existing filter of whether pedestrians are involved in the accident

    This filter overlaps with the collision type filter and the new casualty type filter.

1. Reorder filters

    Filters are reordered to the following order (filter name in `ui.R`):

    1. Date range (`date_filter`)
    2. Severity (`severity_filter`)
    3. Collision type (`collision_type_filter`)
    4. Casualty role (`casualty_filter`)
    5. Vehicles involved (`vehicle_class_filter`)
    6. Number of casualties (`n_causality_filter`)
    
    The order is referenced from [Crash Map](https://www.crashmap.co.uk/Search).

    

***

# Check
- [x] The travis.ci and R CMD checks pass.

***

## Note

- We may discuss ways to enhance and simplify the UI in future development. The filter panel looks a bit long to me now. It takes some time for new users to understand what each filter is doing. 😐

### CODE REVIEW

**Comment by KHwong12** [comment]
[Map Viewer - Start with Pop-up Preview & Full Details Page](https://trello.com/c/tSgikljF/88-map-viewer-start-with-pop-up-preview-full-details-page)

File: inst/app/modules/main_map.R [hunk 1]
```diff
@@ -1,20 +1,21 @@
 
-# Get information of all types of vehicles involved in the accidents
+# Get information of all types of vehicles involved in the accidents to show in popup
 # TODO: prepare following mutated dataset before running shiny to save loading time?
 hk_vehicles_involved = hk_vehicles %>%
```
Comment (avisionh) [inline inst/app/modules/main_map.R:4]: Apply consistent assignment operator.
```suggestion
hk_vehicles_involved <- hk_vehicles %>%
```

File: inst/app/ui.R [hunk 2]
```diff
@@ -131,12 +131,15 @@ ui <- dashboardPage(
...
-                inputId = "n_causality_filter", label = "No. of casualties",
-                min = min(hk_accidents$No__of_Casualties_Injured),
-                max = max(hk_accidents$No__of_Casualties_Injured),
-                value = range(hk_accidents$No__of_Casualties_Injured),
-                step = 1
+              checkboxGroupButtons(
+                inputId = "severity_filter", label = "Accident Severity",
+                choices = c(`Fatal <i style="color:#230B4C;" class="fas fa-skull-crossbones"></i>` = "Fatal",
+                            `Serious <i style="color:#C03A51;"class="fas fa-procedures"></i>` = "Serious",
+                            `Slight <i style="color:#F1701E;" class="fas fa-user-injured"></i>` = "Slight"),
```
Comment (avisionh) [inline inst/app/ui.R:138]: Nice use of icons here.

File: inst/app/ui.R [hunk 3]
```diff
@@ -158,50 +161,50 @@ ui <- dashboardPage(
...
-                justified = TRUE, status = "primary"
+              p("NOTE: Multiple selections means filtering accidents including ANY selected classes
+                (instead of ALL selected classes)."),
+
+              sliderInput(
+                inputId = "n_causality_filter", label = "No. of casualties",
+                min = min(hk_accidents$No__of_Casualties_Injured),
+                max = max(hk_accidents$No__of_Casualties_Injured),
+                value = range(hk_accidents$No__of_Casualties_Injured),
+                step = 1
```
Comment (avisionh) [inline inst/app/ui.R:207]: Should the range of 1 to 77 update according to the filters chosen by the user? If yes, it didn't work when I tried.

**Review by avisionh — CHANGES_REQUESTED** [review]
Wonderful stuff! I've made a few minor comments about using consistent assignment operators and improving the grammar.

There are two other things, outlined below, however, they can be dealt with in a separate PR, which is what you mentioned..
1. Great that you can drag and drop the **Filter Panel** (was pleasantly surprised when I did this by accident). However, it seems to also overflow onto the other parts of the dashboard when you scroll down. (Screenshot provided below).
2. The *No. of casualties* slider doesn't seem to update in accordance with the filters selected. Is this expected behaviour?

image

**Comment by KHwong12** [comment]
Thanks for such a detailed review! Let me address your concerns before looking into the suggestions.

> However, it seems to also overflow onto the other parts of the dashboard when you scroll down. (Screenshot provided below).

You took the words out of my mouth. I also find the panel is now too long and kind of weird. I looked for ways to collapse the panel before (like [this SO question](https://stackoverflow.com/questions/35175167/collapse-absolutepanel-in-shiny)) and would like to give it a go in future development.

Another way we could do is to remove the number of filters? Would it be better to "restrict" the number of choices for users?

> The _No. of casualties_ slider doesn't seem to update in accordance with the filters selected. Is this expected behaviour?

Yep, that's expected behaviour. Like the collision type filter and collision severity filter, the choices available are independent of the users' current selection. It is based on the **whole** `hk_accdients` dataset, where the min and max number of causalities are 1 and 77.

IMO continuously updating the available choices in the filter may be a little bit confusing. We may add some basic summary statistics below the map if the user would like to get the min/max number of causalities in his/her current selection...? 🤔

**Comment by avisionh** [comment]
Thanks for making the tweaks!

> Thanks for such a detailed review! Let me address your concerns before looking into the suggestions.
> 
> > However, it seems to also overflow onto the other parts of the dashboard when you scroll down. (Screenshot provided below).
> 
> You took the words out of my mouth. I also find the panel is now too long and kind of weird. I looked for ways to collapse the panel before (like [this SO question](https://stackoverflow.com/questions/35175167/collapse-absolutepanel-in-shiny)) and would like to give it a go in future development.
> 
> Another way we could do is to remove the number of filters? Would it be better to "restrict" the number of choices for users?

Removing the filters is an interesting suggestion. I do feel there are quite a lot of filters that, from user perspective, can seem overwhelming initially as someone who is not familiar with traffic collisions data. Nevertheless, I remember that the users being targeted with this dashboard are more technical from the angle of knowing more about traffic collisions, so they will benefit from the extra granular detail they can derive from tweaking these filters. In that regard, I think keeping the filters as is would be sufficient. 

I definitely feel being able to collapse the panel will be a great addition in a separate PR.

> 
> > The _No. of casualties_ slider doesn't seem to update in accordance with the filters selected. Is this expected behaviour?
> 
> Yep, that's expected behaviour. Like the collision type filter and collision severity filter, the choices available are independent of users' current selection. It is based on the **whole** `hk_accdients` dataset, where the min and max number of causalities are 1 and 77.
> 
> IMO continuously updating the available choices in the filter may be a little bit confusing. We may add some basic summary statistics below the map if the user would like to get the min/max number of causalities in his/her current selection...? 🤔

Thanks for the explanation, makes sense and I agree with your point about why continuously updating the available choices can be confusing. Filling out those plots and including more summary statistics will be a really great addition.

**Review by avisionh — APPROVED** [review]

**Pull request merged**